### PR TITLE
Add error source

### PR DESCRIPTION
**Pull request by Multimo**
This pr adds a httpclient which based on status code adds ErrorSource to the  allows grafana to know when the error has been internal to the plugin or if its google-sheets itself. 

We could also just import the middleware if thats preferable. 

There are also errors in the `Datasource.QueryData` method. Not sure how you wanna handle them :)

### CODE REVIEW

**Comment by ivanahuckova** [comment]
@Multimo @gabor Could you have a look please. I have pushed https://github.com/grafana/google-sheets-datasource/pull/246/commits/7b919177cf782a0b33bc28fd565b8cf0766da9fa that updates this PR:

1. Instead of using errorsource client, I have replaced it to errorsource middleware. This way, we are not overriding any possible other round trippers and we are just appending error source middleware to the rest of the middlewares.
2. I have marked some of the errors as downstream errors - for example when auth is set up incorrectly or API key is missing.
3. We use middleware for http responses, but we are then doing some custom logic around errors, so I have added adding of source there as well. If source is already set up, it will use that and not override it.  

I have also done a little manual/sanity check on some basic user errors to ensure downstream errors are considered correctly.

https://github.com/grafana/google-sheets-datasource/assets/30407135/240d1403-4d8b-4786-ba09-556785d4b42b

**Comment by ivanahuckova** [comment]
@sasklacz fyi here I am implementing error source in a way where it considers user errors (e.g. not filled API key) as downstream errors.

File: pkg/googlesheets/googlesheets.go [hunk 1]
```diff
@@ -100,18 +102,27 @@ func (gs *GoogleSheets) getSheetData(client client, qm *models.QueryModel) (*she
 	if err != nil {
 		if apiErr, ok := err.(*googleapi.Error); ok {
 			// Handle API-specific errors
+			// We use ErrorSourceFromHTTPStatus to determine error source based on HTTP status code
+			// but if the error has already a source, we use that.
 			if apiErr.Code == 404 {
-				return nil, nil, errors.New("spreadsheet not found")
+				errWithSource := errorsource.SourceError(backend.ErrorSourceFromHTTPStatus(apiErr.Code), errors.New("spreadsheet not found"), false)
```
Comment (gabor) [inline pkg/googlesheets/googlesheets.go:108]: in this branch, i wonder if we need `backend.ErrorSourceFromHTTPStatus`.. i mean, we already know it's a `404`, we know it's the "spreadsheet not found" case.. shouldn't we just hardcode `downstreamerror` ?

File: pkg/googlesheets/googleclient.go [hunk 2]
```diff
@@ -152,7 +155,8 @@ func createDriveService(ctx context.Context, settings models.DatasourceSettings)
```
Comment (Multimo) [inline pkg/googlesheets/googleclient.go:155]: Thanks, I was not sure how to handle the client issues but this makes sense. Just to clarify the empty `settings.AuthenticationType` would be a code issue?

File: pkg/googlesheets/googlesheets.go [hunk 3]
```diff
@@ -100,18 +102,27 @@ func (gs *GoogleSheets) getSheetData(client client, qm *models.QueryModel) (*she
...
 			log.DefaultLogger.Error(apiErr.Error())
-			return nil, nil, errors.New("unknown API error")
+			return nil, nil, errWithSource
 		}
+
 		log.DefaultLogger.Error("unknown error", "err", err)
-		return nil, nil, errors.New("unknown error")
+		// This is an unknown error from the client - if it has error source from error source middleware,
+		// we use that. Otherwise we default to a downstream error.
+		errWithSource := errorsource.DownstreamError(errors.New("unknown error"), false)
```
Comment (gabor) [inline pkg/googlesheets/googlesheets.go:124]: maybe i'm reading this wrong but... to achieve the "if it has error source from error" functionality, we have to propagate the `err` object into `errorsource.DownstreamError`, right? in other words, we cann `errorsource.DownstreamError()` with a fresh new `errors.New()`.. that thing will have no `source`... WDYT?

File: pkg/googlesheets/googlesheets.go (hunk 1 above)
Comment (Multimo) [inline pkg/googlesheets/googlesheets.go:108]: error source has a method to wrap [http status](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/experimental/errorsource/error_source.go#L78)  but it doesnt include those extra params. Potentially we can update the helpers to be more useful eventually.

**Review by Multimo — COMMENTED** [review]
Looks great thanks @ivanahuckova

**Review by gabor — COMMENTED** [review]
i wonder if we "profit" from the errorsource-middleware...it seems to me that the errrors it generates are not propagated back to `QueryData`, but i did not debug this, so maybe i'm just missing the obvoius.

here is how it seems to me:
1. `QueryData` calls into `googlesheets.Query` to create the `DataResponse`.
2. `Query` calls `getSheetData`
3. `getSheetData` calls `client.GetSpreadsheet` at https://github.com/grafana/google-sheets-datasource/blob/58e874329c6fc4b5168ce1602e66a9ef28ffac3f/pkg/googlesheets/googlesheets.go#L101 .. that returns (in case of errors) `err`.. this should be the one coming from the middleware.. i think
4. but, we do not "return" that `err` object in any form from this place.. we check it's attrbutes, but the `err` object is not returned from `getSheetData`,  so it will not reach https://github.com/grafana/google-sheets-datasource/blob/58e874329c6fc4b5168ce1602e66a9ef28ffac3f/pkg/googlesheets/googlesheets.go#L40

WDYT?

File: pkg/googlesheets/googlesheets.go (hunk 3 above)
Comment (ivanahuckova) [reply pkg/googlesheets/googlesheets.go]: Fixed in https://github.com/grafana/google-sheets-datasource/pull/246/commits/f4950d839ee43958f5740226c96ec6fc62abf226 and the 404 error being downstream in https://github.com/grafana/google-sheets-datasource/pull/246/commits/14bc6baeb9509439ed1464f99adc7480343727f0

File: pkg/googlesheets/googleclient.go (hunk 2 above)
Comment (ivanahuckova) [reply pkg/googlesheets/googleclient.go:155]: What do you mean by code issue? Downstream or Plugin?

**Review by gabor — APPROVED** [review]
LGTM 👍

**Pull request merged**